oeps/oep-0007.rst

```diff
@@ -0,0 +1,526 @@
+| Resolution    |                                           |
++---------------+-------------------------------------------+
+
+
+Abstract
+========
+
+With primary development of Python occurring on Python 3, and with Python 2
+scheduled for end-of-life in 2020, edX needs to plan for a transition of all of
+our code to Python 3.  Open edX is a large project, spanning many applications
+over even more github repositories, with even more dependencies on third party
+libraries.  This document outlines how we plan to scope the problem, socialize
+understanding of the differences between versions of Python, migrate our code,
+and work with our Open edX community to ensure that the process is as painless
+as possible, and meets the needs of our stakeholders.
+
+
+Rationale
+=========
+
+    * Python 2 will only receive support from the Python core devs until 2020.
+    * Django will only support Python 2 until 1.11LTS (which is also supported
+      until 2020).
+    * New features are being made available on Python 3 first, and are only
+      sometimes backported to Python 2.
+    * New syntax extensions (``async``/``await``, ``yield from``, and the ``@``
+      operator, for instance) are only being introduced in Python 3.
+    * Python 3 provides better separation between text and binary data.
+    * Python 3 provides better memory usage through a more thorough use of
+      iterators and smart collections.
+    * Major language performance work is being done on Python 3 only.
+    * Almost all major libraries now support Python 3.
+    * We can approach the problem now at a relaxed pace, or deal with it in a
+      panic later, diverting valuable resources from other important work.
+    * The longer we wait, the more code we will produce that needs to be
+      migrated.
+    * Using Python 3 gives us a strong story as a leader in the Python and open
+      source communities.
+    * Python 3 makes for more compelling job postings [Citation needed]
+
+
+Scope
+=====
+
+There are several areas of work that need to be done
+
+* Setting policy
+* Changing coding practice
+* Migrating existing code
+* Deploying edx.org services with Python 3
+* Deprecating Python 2 support
+
+These areas have certain dependencies between them, e.g., we cannot deploy a
+service under Python 3 until all of its code has been migrated.  However, this
+should not be taken as a chronological list of tasks.  We may begin deprecating
+Python 2 support on certain repositories before existing code in other
+repositories has been migrated.
+
+
+Choosing a Python version
++++++++++++++++++++++++++
+
+All new Python projects should be written using Python 3 unless a compelling
+reason (such as incompatible support libraries with no reasonable alternative
+available) necessitates that we stick with Python 2.
+
+
+New libraries should additionally support Python 2 using a single code-base
+with Six_ as a compatibility layer if they will be used by projects that run on
```

Comment:
We have, at times, struggled with different libraries requiring different versions of `six`.